### DISCUSSION TITLE
IMSL_CONSTANT not finding udunits database

### DISCUSSION THREAD

**Discussion by jdtreffert** [gnudatalanguage/gdl#1723] [Q&A]
I'm running V 1.0.4 on Windows and IMSL_CONSTANT is failing with 
% IMSL_CONSTANT: UDUNITS: failed to load the default unit database
Am I missing a file or do require additional configuration?

**Comment by alaingdl** [comment]
I have no MSwin around ... On my laptop running Linux and GDL compiled with udunits, I have :

```
GDL> gdl_status
% Compiled module: GDL_STATUS.
...
21 UDUNITS  :        1
```
and
```
GDL> imsl_constant('amu')
   1.6605387E-27
```

(could you check the GDL_STATUS ?)

You could try to compile GDL in WSL2, it should works fine, I already did that for colleagues months ago

**Comment by alaingdl** [comment]
In fact I llok at the `build_gdl.sh` script and udunits is here for MSwin 

Could you also give a more detailed example ?

**Comment by jdtreffert** [comment]
gdl_status indicates the UDUNITS is present

% Compiled module: GDL_STATUS.

1

DSFMT ? :

.....

21

UDUNITS :

1

However

GDL> print,IMSL_CONSTANT('C')

% IMSL_CONSTANT: UDUNITS: failed to load the default unit database

% Execution halted at: $MAIN$

**Comment by alaingdl** [comment]
OK. I suspect you a problem with the related DLL, but I am totally ignorant with MSwin

Did you run the test suite ? May help to clarify what happen ?

```
GDL> print,IMSL_CONSTANT('C')  
  2.99792e+08
```

**Comment by GillesDuvert** [comment]
Did you install gdl from the installer, or did you compile it? In the first case, it would mean that a needed dll is missing from the installer? (not on MSWin, not even sure that such thing as an udunits.dll exists)

**Comment by slayoo** [comment]
The error is issued from GDL code if UDUNITS XML files cannot be located:
https://github.com/gnudatalanguage/gdl/blob/8157c1239eb8164633b221b5dbd74d5e905b7b3f/src/gsl_fun.cpp#L2881-L2882

So, likely it is not a DLL problem but an incomplete UDUNITS installation that shipped the DLLs, but not the XML files, or some other config issue that prevents UDUNITS from finding the XML files (see https://docs.unidata.ucar.edu/udunits/current/#Database).

**Comment by jdtreffert** [comment]
That sounds right.  I don’t see the XML file(s).

**Comment by alaingdl** [comment]
does that means we have to improve the tests or the message ?

(and yes, we still need a FAQ ! in the Wiki ?)

**Comment by jdtreffert** [comment]
I believe the message is correct as it refers to a missing default database
which is presumably missing XML file(s).  I tried taking XML files from the
UDUNITS  distribution and place in a  GDL path but was not successful.

Do we know the expected XML file and location?

**Comment by slayoo** [comment]
FWIW, here are the locations used in the Debian package: https://packages.debian.org/sid/all/libudunits2-data/filelist
Re how to detect it in CI: indeed we seem to not test it on Windows on CI?

**Reply by GillesDuvert** [reply]
Indeed, there is no way for us to check if a particular installation of a needed library contains all the required ancillary files. Even using the compile-all "build_gdl.sh" procedure, these ancillary files may be contained in other packages (debian for example is known to separate things in small independent packages, knowing that their database keeps the relationship) .
This goes for plplot, that has a dozen of ancillary files not necessary installe with the 'base' package, the proj library,  and probably others.

**Comment by slayoo** [comment]
@jdtreffert how have you installed UDUNITS and GDL?

**Comment by jdtreffert** [comment]
With the windows installer

**Reply by slayoo** [reply]
@GillesDuvert @pjb7687 would this imply that we are not shipping the XML files together with the GDL Windows installer?

**Reply by GillesDuvert** [reply]
of course! As I said, it needs user returns, such as this one, to figure out what is missing in a particular distribution.
One could be confident that everything works with a gdl compiled and installed using 'build_gdl.sh',  because of the test file 'test_constants' working on the windows continuous integration machine, but for the Apple or Windows installer, the files need to be written in the installer package by the "build_gdl.sh pack" command. Obviously these XML files are absent and need to be added.

**Comment by jdtreffert** [comment]
Yes

**Comment by jdtreffert** [comment]
I wanted to let all know that I can resolve this issue on Windows if
following a Windows installation  of v1.04:

1.  Copy the UDUNITS*.XML files from the UDUNITS installation zip to a
directory
2.  Set an environment variable UDUNITS2_XML_PATH to the full path
(directory+name) of the udunits2.xml file

**Reply by GillesDuvert** [reply]
Not clear where to find those XML files.
It is said "This location varies depending upon your operating system and UDUnits2 compilation settings."

**Comment by jdtreffert** [comment]
I am not building the DLL - but I downloaded the .zip distribution file -
https://downloads.unidata.ucar.edu/udunits/2.2.28/udunits-2.2.28.zip
There is a udunits2.xml file in the lib directory of the distribution which
shows dependencies on other udunit*.xml files.
